File: ACTIVIDADES/Actividad N°6/Dijkstra.py
import heapq
import random
import time
from memory_profiler import memory_usage

# Genera un grafo dirigido con pesos aleatorios y conexiones limitadas por nodo
def generar_grafo(n, conexiones_por_nodo=3):
    grafo = {i: [] for i in range(n)}
    for u in range(n):
        conectados = set()
        while len(conectados) < conexiones_por_nodo:
            v = random.randint(0, n - 1)
            if v != u and v not in conectados:
                peso = random.randint(1, 20)
                grafo[u].append((v, peso))
                conectados.add(v)
    return grafo

# Algoritmo de Dijkstra usando min-heap
def dijkstra(graph, start):
    dist = {node: float('inf') for node in graph}
    dist[start] = 0
    priority_queue = [(0, start)]  # (distancia, nodo)

    while priority_queue:
        d, u = heapq.heappop(priority_queue)
        if d > dist[u]:
            continue
        for v, weight in graph.get(u, []):
            if dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight
                heapq.heappush(priority_queue, (dist[v], v))
    return dist

# Evalúa tiempo y memoria para distintos tamaños
def evaluar_dijkstra():
    tamanos = [10, 100, 1000]
    tiempos = []
    memorias = []

    for n in tamanos:
        grafo = generar_grafo(n)

        # Medición de tiempo
        inicio = time.time()
        distancias = dijkstra(grafo, 0)
        fin = time.time()
        duracion = (fin - inicio) * 1000  # ms
        tiempos.append(duracion)

        # Medición de memoria
        uso_memoria = memory_usage((dijkstra, (grafo, 0)), max_iterations=1)
        memoria_max = max(uso_memoria) - min(uso_memoria)
        memorias.append(memoria_max)

        print(f"Grafo de {n} nodos: Tiempo = {duracion:.2f} ms, Memoria ≈ {memoria_max:.2f} MiB")
        print(f"Distancias desde nodo 0 a los primeros 5 nodos: {[distancias[i] for i in range(min(5, n))]}\n")

    return tamanos, tiempos, memorias

# Ejecutar evaluación
if __name__ == "__main__":
    evaluar_dijkstra()

File: ACTIVIDADES/Actividad N°6/Dijkstra.cpp
#include <iostream>
#include <vector>   // Para usar vectores dinámicos
#include <queue>    // Para la cola de prioridad (priority_queue)
#include <limits>   // Para obtener valores límite (infinito)
#include <cstdlib>  // Para números aleatorios (rand, srand)
#include <ctime>
#include <chrono>
#include <unordered_set>    // Para conjuntos hash (evitar duplicados)

using namespace std;

// Función para generar un grafo dirigido con pesos aleatorios y una cantidad fija de conexiones por nodo
vector<vector<pair<int, int>>> generarGrafo(int n, int conexionesPorNodo = 3) {
    vector<vector<pair<int, int>>> grafo(n);
    srand(time(0)); // Semilla aleatoria para obtener diferentes grafos cada vez

    for (int u = 0; u < n; ++u) {
        unordered_set<int> conectados; // Para evitar múltiples conexiones al mismo nodo
        while (conectados.size() < conexionesPorNodo) {
            int v = rand() % n;
            int peso = 1 + rand() % 20; // Peso entre 1 y 20
            if (u != v && conectados.find(v) == conectados.end()) {
                grafo[u].emplace_back(v, peso); // Agrega conexión u -> v
                conectados.insert(v);
            }
        }
    }
    return grafo;
}

// Algoritmo de Dijkstra utilizando un min-heap (priority_queue con greater<>)
vector<int> dijkstra(int n, const vector<vector<pair<int, int>>>& grafo, int inicio) {
    vector<int> dist(n, numeric_limits<int>::max()); // Inicializa distancias a infinito
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq; // Min-heap

    dist[inicio] = 0;
    pq.push({0, inicio}); // (distancia, nodo)

    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();

        if (d > dist[u]) continue; // Si ya hay una mejor distancia, ignorar

        for (const auto& [v, peso] : grafo[u]) {
            if (dist[u] + peso < dist[v]) {
                dist[v] = dist[u] + peso;
                pq.push({dist[v], v}); // Agrega el nuevo candidato al heap
            }
        }
    }

    return dist;
}

int main() {
    // Tamaños del grafo para modelar: 10, 100 y 1000 nodos
    vector<int> tamanos = {10, 100, 1000};

    for (int n : tamanos) {
        auto grafo = generarGrafo(n); // Genera el grafo

        auto inicio = chrono::high_resolution_clock::now(); // Inicia cronómetro

        auto distancias = dijkstra(n, grafo, 0); // Ejecuta Dijkstra desde nodo 0

        auto fin = chrono::high_resolution_clock::now(); // Detiene cronómetro

        auto duracion = chrono::duration_cast<chrono::milliseconds>(fin - inicio).count();

        cout << "Grafo de " << n << " nodos: Tiempo = " << duracion << " ms" << endl;

        // Mostrar un resumen de las distancias para verificar funcionamiento
        cout << "Distancias desde nodo 0 a los primeros 5 nodos: ";
        for (int i = 0; i < min(5, n); ++i) {
            cout << distancias[i] << " ";
        }
        cout << "...\n\n";
    }

    return 0;
}
